src/Std/Internal/Async/Basic.lean

```diff
@@ -109,6 +329,430 @@ def getState (x : AsyncTask α) : BaseIO IO.TaskState :=
+  EAsync.lift <| self.asTask
+
+instance : Functor (EAsync ε) where
+  map := EAsync.map
+
+instance : Monad (EAsync ε) where
+  pure := EAsync.pure
+  bind := EAsync.bind
+
+instance : MonadLift (EIO ε) (EAsync ε) where
+  monadLift := EAsync.lift
+
+instance : MonadExcept ε (EAsync ε) where
+  throw := EAsync.throw
+  tryCatch := EAsync.tryCatch
+
+instance : MonadExceptOf ε (EAsync ε) where
+  throw := EAsync.throw
+  tryCatch := EAsync.tryCatch
+
+instance : MonadFinally (EAsync ε) where
+  tryFinally' := EAsync.tryFinally'
+
+instance : OrElse (EAsync ε α) where
+  orElse := MonadExcept.orElse
+
+instance [Inhabited ε] : Inhabited (EAsync ε α) where
+  default := ⟨pure (MaybeTask.pure default)⟩
+
+instance : MonadAwait (ETask ε) (EAsync ε) where
+  await t := mk <| pure <| .ofTask t
+
+instance : MonadAwait Task (EAsync ε) where
+  await t := mk <| pure <| .ofTask (t.map (.ok))
+
+instance : MonadAwait AsyncTask (EAsync IO.Error) where
+  await t := mk <| pure <| .ofTask t
+
+instance : MonadAwait IO.Promise (EAsync ε) where
+  await t := mk <| pure <| .ofTask (t.result!.map (.ok))
+
+instance : MonadAsync (ETask ε) (EAsync ε) where
+  async t := EAsync.lift <| t.asTask
+
+instance : MonadAsync AsyncTask (EAsync IO.Error) where
+  async t := EAsync.lift <| t.asTask
+
+/--
+A tail recursive version of the `forIn` for while loops inside the `EAsync` Monad.
+-/
+@[inline]
+protected partial def forIn {β : Type} [i : Inhabited ε] (init : β) (f : Unit → β → EAsync ε (ForInStep β)) : EAsync ε β := do
+  let promise ← IO.Promise.new
+
+  let rec @[specialize] loop (b : β) : EAsync ε (ETask ε Unit) := async do
+    match ← f () b with
+      | ForInStep.done b => promise.resolve (.ok b)
+      | ForInStep.yield b => discard <| (loop b)
+
+  discard <| loop init
+
+  .mk <| pure <| .ofTask promise.result!
+
+instance eta [Inhabited ε] : ForIn (EAsync ε) Lean.Loop Unit where
+  forIn _ := EAsync.forIn
+
+end EAsync
+
+/--
+An asynchronous computation that may produce an error of type `IO.Error`.
+Alias for `EAsync IO.Error`.
+-/
+abbrev Async (α : Type) := EAsync IO.Error α
+
+namespace Async
+
+/--
+Converts a `Async` to a `AsyncTask`.
+-/
+@[inline]
+protected def toIO (x : Async α) : IO (AsyncTask α) :=
+  MaybeTask.toTask <$> x.toRawBaseIO
+
+/--
+Returns the `Async` computation inside an `AsyncTask`, so it can be awaited.
+-/
+@[inline]
+def async (self : Async α) : Async (AsyncTask α) :=
+  EAsync.lift <| self.asTask
+
+@[default_instance]
+instance : MonadAsync AsyncTask Async :=
+  inferInstanceAs (MonadAsync (ETask IO.Error) (EAsync IO.Error))
+
+instance : MonadAwait AsyncTask Async :=
+  inferInstanceAs (MonadAwait AsyncTask (EAsync IO.Error))
+
+instance : MonadAwait IO.Promise Async :=
+  inferInstanceAs (MonadAwait IO.Promise (EAsync IO.Error))
+
+end Async
+
+/--
+An asynchronous computation that cannot fail with any error.
+Alias for `EAsync Empty`.
+-/
+abbrev BaseAsync (α : Type) := EAsync Empty α
```

Comment:
I question defining `BaseAsync` in terms of `EAsync` instead of vice versa. I understand that the idea is probably to parallel `BaseIO`'s and `EIO`'s definitions, but the situation is different here. The synchronous monads are defined that way because the return values have to be paired with the `IO.RealWorld`, so merging the pair with the exception is more efficient. For tasks, though, no pairing with `IO.RealWorld` is necessary, so avoiding the unnecessary `Except` in the case of `BaseAsync` is likely more efficient.

Reply:
Now `EAsync` is defined in terms of `BaseAsync`.